File: take_u_forward/graph/36_Minimum Multiplications to reach End.java
// User function Template for Java

class Solution {
    int minimumMultiplications(int[] arr, int start, int end) {
        
        // If start is already equal to end
        if (start == end) return 0;
        
        int mod = 100000;
        
        // Distance array to store minimum steps to reach each number
        int[] dist = new int[mod];
        Arrays.fill(dist, (int) 1e9);
        dist[start] = 0;
        
        // Queue for BFS: {current value, steps}
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{start, 0});
        
        // BFS traversal
        while (!q.isEmpty()) {
            int[] curr = q.poll();
            int node = curr[0];
            int steps = curr[1];
            
            for (int factor : arr) {
                int num = (node * factor) % mod;
                
                // If we found a shorter path
                if (steps + 1 < dist[num]) {
                    dist[num] = steps + 1;
                    
                    // If end is reached
                    if (num == end) {
                        return steps + 1;
                    }
                    
                    q.offer(new int[]{num, steps + 1});
                }
            }
        }
        
        // If end is not reachable
        return -1;
    }
}

File: take_u_forward/graph/36_Minimum Multiplications to reach End.cpp
// User function Template for C++

class Solution {
  public:
    int minimumMultiplications(vector<int>& arr, int start, int end) {
        
        // If start is already equal to end
        if(start == end) return 0;
        
        const int mod = 100000;
        
        // Distance array to store minimum steps to reach each number
        vector<int> dist(mod, 1e9);
        
        // BFS queue: {current value, steps}
        queue<pair<int,int>> q;
        
        q.push({start, 0});
        dist[start] = 0;
        
        while(!q.empty()) {
            int node = q.front().first;
            int steps = q.front().second;
            q.pop();
            
            for(int it : arr) {
                int num = (node * it) % mod;
                
                // If we found a shorter path to num
                if(steps + 1 < dist[num]) {
                    dist[num] = steps + 1;
                    
                    // If end is reached, return answer
                    if(num == end) {
                        return steps + 1;
                    }
                    
                    q.push({num, steps + 1});
                }
            }
        }
        
        // If end is not reachable
        return -1;
    }
};
